### PR TITLE
Param Search

### PR DESCRIPTION
**Pull request by aryamansriram**
This branch consists of 

- preprocess_utils.py: Script containing code reuse frequently
- params_file.json: Parameters file for the CV file
- params_search.py: Script that searches from parameters and applies the same to the model

### CODE REVIEW

File: param_file.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "Model Params",
+  "nParams": 5,
+  "loss": ["binary_crossentropy","binary_crossentropy","binary_crossentropy","binary_crossentropy","binary_crossentropy"],
+  "lr": [1e-5,1e-7,1e-9,1e-10],
+  "epsilon": [1e-8,1e-5,1e-4,1e-9],
+  "epochs": [5,5,5,10],
+  "batch_size": [8,16,16,8]
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply param_file.json]: Done

File: preprocess_utils.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+import json
+import numpy as np
+import pandas as pd
+import pickle
+from colorama import Fore, Style
+from sklearn.model_selection import train_test_split
+import glob
+
+
+def read_file(filepath):
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply preprocess_utils.py]: Done

File: params_search.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        CLASS_WEIGHT_0 = 1
+        CLASS_WEIGHT_1 = (1 - TRAIN_TARGET_POSITIVE_PERCENTAGE) / TRAIN_TARGET_POSITIVE_PERCENTAGE
+
+    MODEL = load_model(CONFIG_DATA)
+
+    for i in range(params_file["nParams"]):
+        MODEL.compile(loss=params_file["loss"][i],
+                      optimizer=Adam(lr=params_file["lr"][i], epsilon=params_file["epsilon"][i]),
+                      metrics=['accuracy'])
+        MODEL.fit(CLF2_TRAIN, CLF2_TRAIN_TARGET,
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply params_search.py:138]: Done

File: param_file.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "Model Params",
+  "nParams": 5,
+  "loss": ["binary_crossentropy","binary_crossentropy","binary_crossentropy","binary_crossentropy","binary_crossentropy"],
+  "lr": [1e-5,1e-7,1e-9,1e-10],
```
[the remark this answers is not in the document]
Comment (aryamansriram) [reply param_file.json]: Done

File: params_search.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
+import argparse
+import numpy as np
+import pandas as pd
+import random
+import tensorflow as tf
+from preprocess_utils import read_file, get_select_vector, import_dataframes
```
Comment (nandan-sridhar) [inline params_search.py:6]:
```suggestion
from preprocess_utils import read_json_file, get_select_vector, import_dataframes
```

File: params_search.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+if __name__ == "__main__":
+    DESCRIPTION = "THIS SCRIPT TAKES A CONFIG AND PARAM FILE AS INPUT AND GENERATES RESULTS FOR ALL PARAMS"
+    ARGUMENT_PARSER = argparse.ArgumentParser(description=DESCRIPTION)
+    ARGUMENT_PARSER.add_argument("-params_path", "--params_path", help="Path to params file", required=True)
+    ARGUMENT_PARSER.add_argument("-config_path", "--config_path", help="Path to config file", required=True)
+
+    PARSED_ARGS = ARGUMENT_PARSER.parse_args()
+    CONFIG_DATA = read_file(PARSED_ARGS.config_path)
+    params_file = read_file(PARSED_ARGS.params_path)
```
Comment (nandan-sridhar) [inline params_search.py:56]:
```suggestion
    CONFIG_DATA = read_json_file(PARSED_ARGS.config_path)
    params_file = read_json_file(PARSED_ARGS.params_path)
```

File: param_file.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      "loss": "binary_crossentropy",
+      "lr": 1e-9,
+      "epsilon": 1e-4,
+      "epochs": 5,
+      "batch_size": 16,
+      "networkCfgJson": "model_configs/binary_relevance_model/topology/Conv1D500-128x128_Dense500_Dense500_Dense500_Dense1_MaxPool.json"
+    }
+  ]
+
+}
```
Comment (nandan-sridhar) [inline param_file.json:31]:
```suggestion
}

```

**Review by nandan-sridhar — COMMENTED** [review]
Please move these files to the models/ folder and update the README.md in the folder with example description.

File: models/params_search.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+import tensorflow as tf
+from preprocess_utils import read_json_file, get_select_vector, import_dataframes
+from sklearn.metrics import classification_report
+from tensorflow.compat.v1.keras.layers import Dense, Conv1D, MaxPooling1D, Flatten
+from tensorflow.compat.v1.keras.models import model_from_json, Sequential
+from tensorflow.compat.v1.keras.optimizers import Adam
+from youtube_audioset import get_recursive_sound_names
+
+params_path = "param_file.json"
+config_path = "model_configs/binary_relevance_model/binary_relevance_model_Wild_cfg.json"
```
Comment (nandan-sridhar) [inline models/params_search.py:14]: Unused variables
```suggestion
```

File: models/README.md [hunk 9]
```diff
@@ -76,3 +76,14 @@ $ python -m models.binary_relevance_model [-h]  -model_cfg_json MODEL_CONFIG_JSO
...
+- This file takes parameters from the ```params_file.json``` file as input and prints out results after training the model using each of them.
+- A variety of hyperparameters can be tested together 
+- The hyperparameters supported include:<br>
+a. loss: The loss function<br>
+b. lr: The learning rate<br>
+c. epsilon: Momentum parameter<br>
+d. batch_size: Batch size<br>
+e. networkCfgJson: Network architecture json file<br>
+- The script can be run using the following command:
+```python models/param_search.py -params_path <Path to the parameter json file> -config_path <Path to the model config file>```
```
Comment (nandan-sridhar) [inline models/README.md:89]:
```suggestion
```python -m models.param_search -params_path <Path to the parameter json file> -config_path <Path to the model config file>```
```

File: models/params_search.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
+import argparse
+import numpy as np
+import pandas as pd
+import random
+import tensorflow as tf
+from preprocess_utils import read_json_file, get_select_vector, import_dataframes
```
Comment (nandan-sridhar) [inline models/params_search.py:6]:
```suggestion
from .preprocess_utils import read_json_file, get_select_vector, import_dataframes
```

**Pull request merged**